### DISCUSSION TITLE
Best way to set some arbitrary data on a text selection?

### DISCUSSION THREAD

**Discussion by JeffDemanche** [facebook/lexical#5273] [Q&A]
Hello all,

TLDR: I'm hoping to hear what approach others would take, as there isn't great guidance on how to implement custom functionality like this. Broadly, I want to append data in HTML attributes to a selection, and wonder if maybe I'm overcomplicating it.

My goal is to allow a user to create a `RangeSelection` with one or more `highlightIds` (represented in the DOM with a comma-separated `data-highlight-ids` value). If `highlightIds` is already present for a node, I want to append to it.

I've tried implementing this by creating a new node that extends `ElementNode` and another that extends `TextNode`, and have run into problems with both approaches.

- With the `ElementNode` approach I'm using a `<span>` with the `data` attribute to wrap text nodes. I find that:
  - I can't apply block-level formatting using `$setBlocksType` to text that contains the new element node.
  - Inserting new lines before, after, or in the middle of text with the new `ElementNode` variously creates `<br>` tags rather than new block-level elements. I've tried overriding the `insert*` methods to limited success.
- With the `TextNode` approach I'm inheriting the DOM element type from the `TextNode` class and adding the attribute on the DOM lifecycle methods, however:
  - With this, block-level formatting works out of the box, but I can't apply inline formatting. My custom text nodes are being converted into vanilla text nodes when I format them. Likewise when I apply my custom text nodes to formatted text, it wipes out the formatting.
 
thanks!
